packages/synapse-react-client/src/components/SchemaDrivenAnnotationEditor/AnnotationEditorUtils.test.ts

```diff
@@ -180,57 +154,276 @@ describe('AnnotationEditorUtils tests', () => {
+          type: 'array',
+          items: {
+            type: 'string',
+          },
+        },
+      })
+    })
+    test('merges properties from oneOf and definitions', () => {
+      const schema: JSONSchema7 = {
+        definitions: {
+          FooDefinition: {
+            type: 'object',
+            properties: {
+              a: {
+                type: 'string',
+              },
+            },
+          },
+        },
+        properties: {
+          b: {
+            type: 'string',
+          },
+        },
+
+        oneOf: [
+          {
+            $ref: '#/definitions/FooDefinition',
+          },
+          {
+            properties: {
+              c: {
+                type: 'array',
+                items: {
+                  type: 'string',
+                },
+              },
+            },
+          },
+        ],
+      }
+
+      const result = getPossibleTopLevelPropertiesInObjectSchema(schema)
+
+      expect(result).toEqual({
+        a: {
+          type: 'string',
+        },
+        b: {
+          type: 'string',
+        },
+        c: {
+          type: 'array',
+          items: {
+            type: 'string',
+          },
+        },
+      })
+    })
+
+    test('override inherited properties when re-defined', () => {
+      const schema: JSONSchema7 = {
+        definitions: {
+          FooDefinition: {
+            type: 'object',
+            properties: {
+              a: {
+                type: 'number',
+              },
+            },
+          },
+        },
+        properties: {
+          a: {
+            type: 'string',
+          },
+        },
+        allOf: [
+          {
+            $ref: '#/definitions/FooDefinition',
+          },
+        ],
+      }
+
+      const result = getPossibleTopLevelPropertiesInObjectSchema(schema)
+
+      expect(result).toEqual({
+        a: {
+          type: 'string',
+        },
+      })
+    })
+
+    test('handle conditional logic', () => {
+      const schema: JSONSchema7 = {
+        properties: {
+          a: {
+            type: 'string',
+          },
+        },
+        if: {
           properties: {
             a: {
+              const: 'foo',
+            },
+          },
+        },
+        then: {
+          properties: {
+            b: {
               type: 'string',
             },
           },
         },
-      },
-      properties: {
+        else: {
+          properties: {
+            c: { type: 'string' },
+          },
+        },
+      }
+
+      const result = getPossibleTopLevelPropertiesInObjectSchema(schema)
+
+      expect(result).toEqual({
+        a: {
+          type: 'string',
+        },
         b: {
           type: 'string',
         },
-      },
-
-      allOf: [
-        {
-          $ref: '#/definitions/FooDefinition',
+        c: {
+          type: 'string',
         },
-        {
+      })
+    })
+
+    test.failing(
+      'ignores properties in definitions that are not top-level',
```

Comment:
Add a failing test for a known bug. I don't know that this bug causes any problems today (planning to do more research), but I wanted to document this issue so we know it exists.

To fix this, we would have to write a utility to walk through the schema tree and resolve references.